Userland/Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -3857,4 +3935,126 @@ Vector<JS::NonnullGCPtr<Element>> Document::elements_from_point(double x, double
     return sequence;
 }
 
+// https://html.spec.whatwg.org/multipage/dom.html#exposed
+static bool is_exposed(Element const& element)
+{
+    if (is<HTML::HTMLObjectElement>(element) || is<HTML::HTMLEmbedElement>(element)) {
+        // FIXME: An embed or object element is said to be exposed if it has no exposed object ancestor, and,
+        //        for object elements, is additionally either not showing its fallback content or has no object or embed descendants.
+        return true;
+    }
+
+    return true;
+}
+
+// https://html.spec.whatwg.org/multipage/dom.html#dom-document-namedItem-which
+Vector<FlyString> Document::supported_property_names() const
+{
+    //  The supported property names of a Document object document at any moment consist of the following,
+    //  in tree order according to the element that contributed them, ignoring later duplicates,
+    //  and with values from id attributes coming before values from name attributes when the same element contributes both:
+    OrderedHashTable<FlyString> names;
+
+    // the value of the name content attribute for all exposed embed, form, iframe, img, and exposed object elements
+    // that have a non-empty name content attribute and are in a document tree with document as their root;
+
+    // the value of the id content attribute for all exposed object elements that have a non-empty id content attribute
+    // and are in a document tree with document as their root; and
+
+    // the value of the id content attribute for all img elements that have both a non-empty id content attribute
+    // and a non-empty name content attribute, and are in a document tree with document as their root.
+
+    for (auto const& element : m_potentially_named_elements) {
+        if (!is_exposed(element))
+            continue;
+
+        if (is<HTML::HTMLObjectElement>(*element)) {
+            if (auto id = element->id(); id.has_value())
+                names.set(id.value());
+        }
+        if (is<HTML::HTMLImageElement>(*element)) {
+            auto maybe_name = element->name();
+            // Only set id if both name and id have value, for img elements. clear as mud
+            if (auto maybe_id = element->id(); maybe_name.has_value() && maybe_id.has_value())
+                names.set(maybe_id.value());
+        }
+
+        if (auto name = element->name(); name.has_value())
+            names.set(name.value());
+    }
+
+    Vector<FlyString> result;
+    result.ensure_capacity(names.size());
+    for (auto const& name : names)
+        result.unchecked_append(name);
+
+    return result;
```

Comment:
`return names.values();` :)

Reply:
woah. There's a few other places we could use that instead of this pattern...